Refactor : 웹 애플리케이션 인증 및 권한 관리 리팩토링

newnyee opened the pull request:
## 🛠 리팩토링 이유
- ### **클라이언트 코드에 의존적**
  AccessToken을 LocalStorage에 저장되어 서버에서 토큰에 접근할 수 없는 문제가 발생함
따라서 매 페이지로 접속 시 Ajax를 이용하여 토큰 검증 API를 호출해야함

- ### **클라이언트 사이드 렌더링(CSR) 방식의 코드**
  프로젝트가 서버사이드 렌더링(SSR) 방식으로 진행되어야 하지만, Ajax를 통한 토큰 검증으로 인해 Thymeleaf를 사용한 렌더링을 할 수 없는 상황 발생

- ### **하나의 인터셉터로 관리**
  하나의 인터셉터로 모든 처리를 하다 보니 코드가 길어지고, 관심사의 분리가 되지 않는 문제가 있음

- ### **LocalStorage에 회원 정보 저장**
  페이지마다 지속적으로 사용될 값(타입, 이미지, 닉네임)을 토큰 갱신시 LocalStorage에 저장하다보니 동시성 문제 발생
  1. 일반 회원이 호스트로 전환할 경우 LocalStorage 회원 정보가 갱신되지 않음
  2. 여러 컴퓨터 또는 브라우저에 로그인이 되어 있는 경우, 갱신된 회원 정보가 반영되지 않음
<br>

## 💻 작업 내용
- ### **Access Token을 쿠키에 저장** 024ccf15c68cecca3110e51d2f0b770ad9e819b3
  서버에서도 사용자의 Access Token에 접근 가능 → 클라이언트에서 Token을 받지 않아도 되면서 의존도가 낮아지며 서버 사이드 렌더링(SSR) 방식을 적용할 수 있게됨

- ### **인터셉터 분리** 229da844350a1f1d011c959b048f69e57b4112f3
  관리와 유지보수를 쉽게 하고 관심사를 분리하기 위해 다음과 같이 총 4개의 인터셉터를 분리함
    
  ###  View URI
    > ### 인증
    > 비회원 상태인 경우에만 접근 가능 페이지
    >- 토큰이 없는 경우 : 그대로 컨트롤러 진입
    >- 토큰이 있는 경우 : 홈으로 리다이렉트
    > <br>
    >
    > 모두가 접근 가능한 페이지
    >- 토큰이 없는 경우 : 그대로 컨트롤러 진입
    >- 토큰이 있는 경우 : 토큰 검증 후 해당 컨트롤러 진입, 클레임 아이디를 request에 저장
    > <br>
    >
    >회원만 접근 가능한 페이지
    >- 토큰이 없는 경우 : 로그인 페이지로 리다이랙트, 현재 요청한 url을 request에 저장
    >- 토큰이 있는 경우 : 토큰 검증 후 컨트롤러 진입, 클레임 아이디를 request에 저장

    > ### 인가
    > - 일반 회원인 경우 : 호스트 가입 페이지로 리다이랙트
    > - 호스트인 경우 : 클레임 아이디를 request에 저장

  ### Api URI
    > ### 인증
    > - 토큰이 있는 경우 : 컨트롤러 진입
    > - 만료된 토큰 : session 재발급 후 컨트롤러 진입
    > - 토큰이 없는 경우 & 유효하지 않은 토큰 : response status를 UNAUTHORIZED로 setting, return false
  
    > ### 인가
    > - 일반 회원인 경우 : response status를 FORBIDDEN으로 setting, return false
    > - 호스트인 경우 : 컨트롤러 진입

- ### **회원 정보 렌더링을 위한 인터셉터의 postHandle() 메서드 활용** [`a7a84e1`](https://github.com/newnyee/refactoring-habit/pull/19/files#diff-a7a84e18d01ffa33011c78872e598a7aa3201448e32662e3e6ebdb83dfba212a)
   - 이 인터셉터는 `뷰 컨트롤러를 호출`하며, `세션을 유지하고 있는 회원`에게만 동작함
   - `세션을 유지하고 있는 회원` : preHandle() 메서드에서 `회원일 경우 request에 “altId”를 저장`하므로 “altId”를 이용하여 회원 여부를 판단하고, 회원일 경우 model에 회원 정보를 저장함
<br>

## 🔎 PR 특이 사항
- AuthServiceTest 코드 변경 03d83d573133fdc2a394504e78cd1f98cd96a2cf
  - Tests passed : 8 of 8
![image](https://github.com/newnyee/refactoring-habit/assets/121937711/b9c358ab-1c7b-44c3-9e1f-6c2d453d9dcd)

- Interceptor test 코드 작성 fda3d9266a9a27f65c9ddb836c4ab6f75d517ecb
  - Tests passed : 20 of 20
![image](https://github.com/newnyee/refactoring-habit/assets/121937711/f9f1b3c7-bbee-43f1-9ec2-39b88ce180fe)

newnyee committed:
refactor: token Response 형태 수정

- 사용되지 않는 필드 제거 (토큰 만료시간)

newnyee committed:
refactor: 모든 토큰 Cookie 저장으로 로직 변경

- 서버에서 토큰에 직접 접근이 가능하도록 Access, Refresh 토큰 모두 Cookie 저장

newnyee committed:
refactor: CookieUtil.getTokenInCookie 리팩토링

- 재사용성을 높이기 네이밍 변경, 매개변수 추가

newnyee committed:
refactor: AuthServiceTest 리팩토링

- 유틸 코드 변경으로인한 test 코드 수정
- 중복 코드 제거를 위한 리팩토링

newnyee committed:
refactor: TokenUtil 리팩토링

- 단일 책임 원칙에 의거한 코드 리팩토링
- getTokenNumber() 메서드 추가 : 접두사를 제거한 토큰을 얻을 수 있는 메서드

newnyee committed:
feat: enum 클래스 생성

- AttributeNames : 값 저장시 사용될 네이밍 모음
- UrlMappings : url String 모음

newnyee committed:
feat: ErrorType 추가

newnyee committed:
feat: InterceptorUtils 생성

- 코드 중복을 줄이기 위한 Util 클래스 생성

newnyee committed:
refactor : 인터셉터 분리

- 관리와 유지보수를 쉽게 하고 관심사를 분리하기 위한 인터셉터 분리
- 비회원만 접근, 모두 접근, 회원만 접근, 호스트만 접근으로 분리

newnyee committed:
feat : 회원 정보 렌더링을 위한 인터셉터 추가

- 뷰 컨트롤러를 호출하고 세션을 유지하고 있는 회원에게만 동작함
- 해당 회원의 정보를 model에 저장

newnyee committed:
refactor : WebConfig 수정

- 요구사항에 맞게 Path 분리하여 설정함

newnyee committed:
test : 인터셉터 test 코드 작성

- Tests passed : 19 of 19

newnyee committed:
refactor: 사용되지 않는 코드 삭제

- 쿠키에 토큰 저장으로 인해 토큰 검증 api 호출 코드가 사용되지 않으므로 삭제함

taejin-chang commented on src/main/java/com/refactoringhabit/common/utils/TokenUtil.java:
```diff
@@ -31,13 +34,25 @@ public TokenResponse createToken(String altId) {
         return decodeToken.getClaim(CLAIM_MEMBER_ID).asString();
     }
+
+    public String getTokenNumber(String token) {
+        String[] accessTokenParts = token.split(" ");
+        if (accessTokenParts.length != 2 || !PREFIX_TOKEN.equals(accessTokenParts[0])) {
+            throw new InvalidTokenException();
+        }
+        return accessTokenParts[1];
+    }
```
token을 만들때 Bearer prefix를 사용하고 있지 않은 것 같습니다. 
그렇다면 해당 메서드는 제거해도 되지 않을까요?

taejin-chang commented on src/main/java/com/refactoringhabit/auth/domain/service/AuthService.java:
```diff
@@ -76,22 +78,18 @@ public SignInResponseDto authenticationAndCreateToken(
+                response, REFRESH_TOKEN_COOKIE_NAME, createdRefreshToken);
+            cookieUtil.createTokenCookie(
+                response, ACCESS_TOKEN_COOKIE_NAME, createdToken.accessToken());
+        } else {
+            throw new UserNotFoundException();
```
PasswordNotMatchException 같은 네이밍으로 별도의 exception을 사용하는 것이 조금 더 적합할 것 같다고 생각됩니다.

taejin-chang commented on src/main/java/com/refactoringhabit/common/enums/AttributeNames.java:
```diff
@@ -0,0 +1,18 @@
+
+public enum AttributeNames {
+    
+    ALT_ID("altId"),
```
altId가 모든 테이블에서 대체키로 사용되고 있는 컬럼명이라서, 여기저기에서 사용될 경우 어떤테이블의 altId가 사용되는지 class명과 사용되는 코드를 보면서 추측해야하는 것 같아요. 

차라리 memberAltId 같은 이름으로 사용한다면 길긴하지만 직관적으로 받아들일 수 있을것 같아요.

변경에 대해 확인해주시면 좋을것 같습니다.

ParkHaeBeen commented:
관심사 분리를 위해 인터셉터를 여러개 분리할 수도 있지만
어노테이션과 인터셉터를 활용해서 조금 합쳐보면 어떨까요??

현재 만드신 인터셉터를 정리하면 토큰이 있는지, 회원/호스트인지에 따라 나뉜것 같은데

```
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface Auth {

  boolean required() default true;
  boolean isAdmin() default false;
}

```

```
@RequiredArgsConstructor
@Component
public class AuthInterceptor implements HandlerInterceptor {

  private final JwtTokenProvider jwtTokenProvider;

  @Override
  public boolean preHandle(final HttpServletRequest request ,final HttpServletResponse response ,final Object handler) throws Exception {
    final Auth authAnnotation = getLoginAnnotation(handler);

    if(!(handler instanceof HandlerMethod) || authAnnotation ==null){
      return true;
    }

    if(isAuthExist(request)){
      validateToken(request);
      isAdmin(request, authAnnotation);
      return true;
    }

    validateTokenRequired(authAnnotation);
    return true;
  }
```

이런 느낌으로 짜도 괜찮을 것같습니다.
(제코드의 일부를 가져온거라서 private메서드는 생략했습니다)

newnyee commented:
@ParkHaeBeen 말씀하신 부분 어떤 의미인지 이해했습니다. 
저도 처음에는 올려주신 코드처럼 하나의 인터셉터로 관리해보려했으나 페이지 마다의 특성이 다르기때문에(같은 조건이어도 로직이 달라짐) 위의 4개의 인터셉터로 분리되었습니다.
좀 더 고민해보고 의견 반영해보겠습니다.
좋은 의견 감사합니다.

newnyee committed:
refactor: exception 변경

- 비밀번호 매칭 실패 예외이므로 PasswordNotMatchException 으로 변경

newnyee committed:
style: attribute name 변경

- altId -> 모든 테이블 마다 존재하는 컬럼명이므로 보다 명확하고 직관적으로 이해할 수 있도록 memberAltId로 변경

newnyee committed:
refactor: 쿠키 저장 로직 리팩토링

- createTokenCookie → createSessionCookie로 변경
- 저장될 데이터 타입 Object로 변경, 직렬화하여 저장

taejin-chang reviewed (approved):
LGTM!!!

newnyee committed:
fix: token prefix 검증 로직 삭제

- 사용되지 않는 토큰 접두사 검증 로직 삭제

newnyee committed:
refactor: 이메일 중복 확인 api url 변경

- member 경로에서 auth 경로로 변경 (api/v2/auth/check-email)

newnyee committed:
refactor: Interceptor 분리

- 기존의 분리 체계 : AccessLevel로 분리
- 분리 체계 리팩토링 : api와 view로 분리

newnyee committed:
feat: config 추가

- 새로운 인터셉터를 위한 config 추가 (ApiConfig, ViewConfig)

newnyee committed:
feat: enum 추가, 수정

- UriAccessLevel : 인터셉터에서 사용될 enum 추가
- UriMappings : 네이밍 변경

newnyee committed:
feat: 인터셉터 test 코드 변경, 추가

- Tests passed: 20 of 20 tests

newnyee committed:
refactor: 사용하지 않는 파일 삭제

newnyee committed:
refactor: 가독성을 위한 InterceptorUtils 메서드 추가
